datalad/interface/download_url.py

```diff
@@ -35,77 +39,137 @@ class DownloadURL(Interface):
+            yield get_status_dict(
+                status="error",
+                message=(
+                    "When specifying multiple urls, --path should point to "
+                    "an existing directory. Got %r", path),
+                type="file",
+                path=path,
+                **common_report)
+            return
         if not path:
-            path = curdir
+            path = ds.path if ds else curdir
 
         # TODO setup fancy ui.progressbars doing this in parallel and reporting overall progress
         # in % of urls which were already downloaded
         providers = Providers.from_config_files()
-        downloaded_paths, failed_urls = [], []
+        downloaded_paths = []
+        path_urls = {}
         for url in urls:
             # somewhat "ugly"
             # providers.get_provider(url).get_downloader(url).download(url, path=path)
             # for now -- via sugaring
             try:
                 downloaded_path = providers.download(url, path=path, overwrite=overwrite)
-                downloaded_paths.append(downloaded_path)
-                # ui.message("%s -> %s" % (url, downloaded_path))
             except Exception as e:
-                failed_urls.append(url)
-                ui.error(exc_str(e))
-                if stop_on_failure:
-                    break
-        if failed_urls:
-            raise RuntimeError("%d url(s) failed to download" % len(failed_urls))
-        return downloaded_paths
+                yield get_status_dict(
+                    status="error",
+                    message=exc_str(e),
+                    type="file",
+                    path=path,
+                    **common_report)
+            else:
+                downloaded_paths.append(downloaded_path)
+                path_urls[downloaded_path] = url
+                yield get_status_dict(
+                    status="ok",
+                    type="file",
+                    path=downloaded_path,
+                    **common_report)
+
+        if downloaded_paths and save and ds is not None:
+            msg = message or """\
+[DATALAD] Download URLs
+
+path: {}
+url:
+  {}""".format(path, "\n  ".join(urls))
+
+            for r in ds.add(downloaded_paths, message=msg):
+                yield r
+
+            if isinstance(ds.repo, AnnexRepo):
+                annex_paths = [p for p, annexed in
+                               zip(downloaded_paths,
+                                   ds.repo.is_under_annex(downloaded_paths))
+                               if annexed]
+                if annex_paths:
+                    keys = assure_list_from_str(
+                        ds.repo.get_file_key(annex_paths))
+                    for key, path in zip(keys, annex_paths):
+                        if key:
+                            # TODO: Add a registerurl method to annexrepo?
+                            # export_to_figshare also uses it.
```

Comment:
There is `add_url_to_file()`

Reply:
There is `add_url_to_file()`

Reply:
> There is add_url_to_file()

Yes, I'm aware.  I used it in the addurls extension.  When thinking about the request in #2123, I came to the conclusion that the idea here is that we should go through `Providers`.  Perhaps I'm missing something though, in which case hopefully @yarikoptic can clarify.

Reply:
Sorry, looking at that again, I think I missed your point: that, with the content already present, `add_url_to_file`  can be used in place of `registerurl`.  I suppose the downside of that is an extra check over the network when we know we just downloaded that url to the file.  I'd guess using addurl rather than registerurl would fail if the provider is using any authentication, but then gain I suppose it doesn't much sense to register the url in that case either.

Reply:
Fwiw If it is a provider whom we support through datalad special remote, it would handle authentication